omegaconf/grammar/OmegaConfGrammarParser.g4

```diff
@@ -0,0 +1,70 @@
+// Regenerate parser by running 'python setup.py antlr' at project root.
+
+// Maintenance guidelines when modifying this grammar:
+//
+// - For initial testing of the parsing abilities of the modified grammer before
+//   writing all the support visitor code, change the test
+//        `tests/test_interpolation.py::test_all_interpolations`
+//   by setting `dbg_test_access_only = True`, and run it. You will also probably
+//   need to comment / hijack the code accesssing the visitor. Tests that expect
+//   errors raised from the visitor will obviously fail.
+//
+// - Update Hydra's grammar accordingly, and if you added more cases to the test
+//   mentioned above, copy the latest version of `TEST_CONFIG_DATA` to Hydra (see
+//   Hydra's test: `tests/test_overrides_parser.py::test_omegaconf_interpolations`).
+
+// - Keep up-to-date the comments in the visitor (in `grammar_visitor.py`)
+//   that contain grammar excerpts (within each `visit...()` method).
+//
+// - Remember to update the documentation (including the tutorial notebook)
+
+parser grammar OmegaConfGrammarParser;
+options {tokenVocab = OmegaConfGrammarLexer;}
+
+// Main rules used to parse OmegaConf strings.
+
+configValue: toplevel EOF;
+singleElement: element EOF;
+
+// Top-level: strings (that need not be parsed), potentially mixed with interpolations.
+
+toplevel: toplevelStr | (toplevelStr? (interpolation toplevelStr?)+);
+toplevelStr: (ESC | ESC_INTER | TOP_CHAR | TOP_STR)+;
```

Comment:
Should this be a lexer token?

Reply:
No, the lexer works greedily, so it would consume everything in a single token if we are not careful. For instance if we had a lexer token defined as `(ESC | ESC_INTER | TOP_CHAR | TOP_STR)+` then the string `abc${foo}` would become a single token (obtained by the concatenation of a `TOP_STR`, `TOP_CHAR` and another `TOP_STR`)